### PR TITLE
[Core] Allow locking/unlocking of `MutexLock`

### PR DESCRIPTION
**Pull request by AThousandShips**
Enables more safe use of RAII with mutexes, fixes some edge cases

### CODE REVIEW

File: core/io/resource_loader.cpp [hunk 1]
```diff
@@ -633,9 +633,9 @@ Ref<Resource> ResourceLoader::load_threaded_get(const String &p_path, Error *r_e
 					// This local poll loop is not needed.
 					break;
 				}
-				thread_load_lock.~MutexLock();
+				thread_load_lock.unlock();
 				OS::get_singleton()->delay_usec(1000);
-				new (&thread_load_lock) MutexLock(thread_load_mutex);
+				thread_load_lock.lock();
```
Comment (AThousandShips) [inline core/io/resource_loader.cpp:638]: This part was the primary incentive now, where the explicit usage was presented

File: core/io/resource_loader.cpp [hunk 2]
```diff
@@ -691,7 +691,7 @@ Ref<Resource> ResourceLoader::_load_complete_inner(LoadToken &p_load_token, Erro
 			Error wtp_task_err = FAILED;
 			if (loader_is_wtp) {
 				// Loading thread is in the worker pool.
-				thread_load_mutex.unlock();
+				p_thread_load_lock.unlock();
```
Comment (AThousandShips) [inline core/io/resource_loader.cpp:694]: This helps with safety as this would have the lock be locked *and* the mutex itself unlocked, this ensures the mutex isn't operated on while the lock is existing

File: core/os/safe_binary_mutex.h [hunk 3]
```diff
@@ -93,16 +93,44 @@ template <int Tag>
...
 public:
 	_ALWAYS_INLINE_ explicit MutexLock(const SafeBinaryMutex<Tag> &p_mutex) :
-			lock(p_mutex.mutex) {
+			_lock(p_mutex.mutex) {
 		SafeBinaryMutex<Tag>::count++;
-	};
+	}
+
 	_ALWAYS_INLINE_ ~MutexLock() {
 		SafeBinaryMutex<Tag>::count--;
```
Comment (kus04e4ek) [inline core/os/safe_binary_mutex.h:105]: This implies that `MutexLock` must be locked when it hits the destructor, is this intentional? If yes, I think there should be a note somewhere about it and `DEV_ASSERT(SafeBinaryMutex<Tag>::count)` or something like this in the destructor

File: core/os/safe_binary_mutex.h [hunk 4]
```diff
@@ -93,16 +93,44 @@ template <int Tag>
...
+			return true;
+		} else {
+			if (_lock.try_lock()) {
+				SafeBinaryMutex<Tag>::count++;
+				return true;
+			} else {
+				return false;
+			}
+		}
+	}
```
Comment (AThousandShips) [inline core/os/safe_binary_mutex.h:133]: This is taken directly from the mutex class, it should be equivalent but it's not used anywhere so hard to confirm

I added the try methods for parity with the mutexes, can remove if desired

File: core/io/resource_loader.cpp (hunk 2 above)
Comment (kus04e4ek) [reply core/io/resource_loader.cpp:694]: Do changes here account for this: https://github.com/godotengine/godot/pull/93739#discussion_r1659827245?

File: core/os/safe_binary_mutex.h (hunk 3 above)
Comment (AThousandShips) [reply core/os/safe_binary_mutex.h:105]: There is, in the Mutex destructor, the lock isn't responsible for this

File: core/io/resource_loader.cpp (hunk 2 above)
Comment (AThousandShips) [reply core/io/resource_loader.cpp:694]: The unlock is equivalent so there's no difference, and see below

File: core/os/safe_binary_mutex.h (hunk 3 above)
Comment (AThousandShips) [reply core/os/safe_binary_mutex.h:105]: But regardless this PR didn't introduce this particular code so I think if there's some issue with the structure here it'd be best to open an issue and make a dedicated fix/improvement

Comment (kus04e4ek) [reply core/os/safe_binary_mutex.h:105]: > There is, in the Mutex destructor, the lock isn't responsible for this, it's only when the Mutex is destroyed that this should be checked

I meant `DEV_ASSERT(SafeBinaryMutex<Tag>::count)` before `SafeBinaryMutex<Tag>::count--`, so there wouldn't be negative numbers

Comment (AThousandShips) [reply core/os/safe_binary_mutex.h:105]: That's a good point, will look

Comment (AThousandShips) [reply core/os/safe_binary_mutex.h:105]: Added that and an additional improvement as a separate commit

File: core/os/safe_binary_mutex.h [hunk 5]
```diff
@@ -93,16 +93,50 @@ template <int Tag>
 class MutexLock<SafeBinaryMutex<Tag>> {
 	friend class ConditionVariable;
 
-	THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> lock;
+	mutable THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> _lock;
 
 public:
+	// Defer locking to ensure locking is done correctly.
```
Comment (AThousandShips) [inline core/os/safe_binary_mutex.h:99]: Can split these changes off into a separate PR if desired, or drop them, just a possible improvement

File: core/os/safe_binary_mutex.h [hunk 6]
```diff
@@ -93,16 +93,50 @@ template <int Tag>
...
+	_ALWAYS_INLINE_ void lock() const {
+		if (++SafeBinaryMutex<Tag>::count == 1) {
+			_lock.lock();
+		}
+	}
+
+	_ALWAYS_INLINE_ void unlock() const {
+		DEV_ASSERT(SafeBinaryMutex<Tag>::count);
+		if (--SafeBinaryMutex<Tag>::count == 0) {
+			_lock.unlock();
```
Comment (AThousandShips) [inline core/os/safe_binary_mutex.h:123]: Might need to add a `owns_mutex` here as well, will see

File: core/os/safe_binary_mutex.h [hunk 7]
```diff
@@ -93,16 +93,51 @@ template <int Tag>
...
+	_ALWAYS_INLINE_ void lock() const {
+		if (++SafeBinaryMutex<Tag>::count == 1) {
+			_lock.lock();
+		}
+	}
+
+	_ALWAYS_INLINE_ void unlock() const {
+		DEV_ASSERT(SafeBinaryMutex<Tag>::count);
+		if (--SafeBinaryMutex<Tag>::count == 0) {
+			DEV_ASSERT(_lock.owns_lock());
```
Comment (AThousandShips) [inline core/os/safe_binary_mutex.h:123]: Just a check to make debugging easier, this will fail with unclear errors in some contexts

**Comment by RandomShaper** [comment]
Overall, I like the new changes. I'd just would like to suggest making the APIs for locking-unlocking a lock more distinct so while skimming through code one knows at a glance if a certain code line is dealing with a mutex or a lock. For instance:
- `temp_unlock()`, `temp_lock()`.
- `delock()`, `relock()`.

**Comment by AThousandShips** [comment]
Will look at some renames, thank you!

**Comment by AThousandShips** [comment]
Changed to `temp_*` to see, will see if I remove the `try_lock` part entirely as it's currently unused and can be implemented again later

File: core/os/safe_binary_mutex.h [hunk 8]
```diff
@@ -93,16 +93,39 @@ template <int Tag>
 class MutexLock<SafeBinaryMutex<Tag>> {
 	friend class ConditionVariable;
 
-	THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> lock;
+	mutable THREADING_NAMESPACE::unique_lock<THREADING_NAMESPACE::mutex> _lock;
```
Comment (RandomShaper) [inline core/os/safe_binary_mutex.h:96]: Despite a few Godot classes are using leading underscore for private "fields," I think the most established rule is to use that only for function names.

**Comment by RandomShaper** [comment]
Looks good to me. My only remaining suggestion would be to rename `temp_lock()` to `temp_relock()` so it's clearer that a `temp_unlock()` must have happened earlier somewhere.

File: core/os/safe_binary_mutex.h (hunk 8 above)
Comment (AThousandShips) [reply core/os/safe_binary_mutex.h:96]: Was only named such because of the name clash with `lock()`, now it's unnecessary so will rename it

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank *you*!

**Comment by AThousandShips** [comment]
Realized I removed some necessary checks when I rewrote the code to adjust to the new Mutex changes, will re-add them to ensure behavior when having unlocked the mutex and it is then destroyed, will push momentarily
